intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp

```diff
@@ -149,56 +166,64 @@ void Model::reinit() {
-          /*.n_prompt_tokens    =*/0,
-          /*.n_past             =*/0,
-          /*.n_total            =*/0,
-          /*.request_idx        =*/bs,
-          /*.beam_idx           =*/0,
-          /*.padding_side       =*/0,
-          /*n_padding           =*/count,
-      });
-    }
-    return post_beam_search(ctx, n_remain, inputs, params.n_threads);
+std::vector<std::vector<model_token>> Model::beam_generate(const std::vector<std::vector<model_token>>& input_ids) {
+  MODEL_ASSERT(input_ids.size() == ctx->batch_size);
+  if (ctx->batch_size > 1 && ctx->vocab.pad_token_id == -1) {
+    fprintf(stderr, "\nERROR: please set pad_token for beam search multi-batch generation!\n");
+    return {{}};
+  }
+  std::vector<model_input> inputs;
+  for (int bs = 0; bs < input_ids.size(); ++bs) {
+    uint32_t count = 0;
+    model_vocab::id pad_token_id = ctx->vocab.pad_token_id;
+    auto iter = std::find_if(input_ids[bs].begin(), input_ids[bs].end(),
+                             [&pad_token_id](model_token t) { return (t != pad_token_id); });
+    if (iter == input_ids[bs].end()) fprintf(stderr, "\nERROR: there are all pad tokens in batch %d!\n", bs);
+    count = std::distance(input_ids[bs].begin(), iter);
+    inputs.push_back(model_input{
+        /*.tokens              =*/input_ids[bs].data(),
+        /*.n_tokens           =*/(uint32_t)input_ids[bs].size(),
+        /*.n_prompt_tokens    =*/0,
+        /*.n_past             =*/0,
+        /*.n_total            =*/0,
+        /*.request_idx        =*/bs,
+        /*.beam_idx           =*/0,
+        /*.padding_side       =*/0,
+        /*n_padding           =*/count,
+    });
   }
+  return post_beam_search(ctx, params.n_predict, inputs, params.n_threads);
+}
+
+const std::vector<float>& Model::evaluate_(const std::vector<std::vector<model_token>>& input_ids) {
+  static const std::vector<float> empty_ret{};
   if (input_ids.size() > 1) {
     fprintf(stderr, "\nERROR: Only beam search supports multi-batch generation!\n");
-    return rets;
+    return empty_ret;
   }
-  if (curr_input_ids.empty()) {
-    if (input_ids[0].size() > n_ctx - 4) {
-      fprintf(stderr, "\n%s: Warning: prompt is too long (%d tokens, max %d), will be truncated\n", __func__,
-              input_ids[0].size(), n_ctx - 4);
-      curr_input_ids.resize(n_ctx - 4);
-      std::copy(input_ids[0].end() - n_ctx - 4, input_ids[0].end(), curr_input_ids.begin());
-    } else {
-      curr_input_ids = input_ids[0];
+
+  const auto& input_id0 = input_ids[0];  // currently only support single batch
+  if (input_id0.empty()) {               // use internel input id
+    if (curr_input_ids.empty()) {
+      fprintf(stderr, "%s: error: no input\n", __func__);
+      return empty_ret;
     }
+  } else if (!curr_input_ids.empty()) {
+    fprintf(stderr, "%s: error: prompt confliction\n", __func__);
+    return empty_ret;
+  } else if (input_id0.size() > n_ctx - 4) {  // long input_id0 and empty curr_input_ids
```

Comment:
Do you think we can move this checking outside from this `evaluate_` function? I also want to use it before beam search generation. Or move this `evaluate_` into `model_utils.cpp`?

Reply:
will be resolved after detailed discussion.